Data Wrangling

Md Tohidul Islam


In [1]:
import numpy as np
import pandas as pd

import pandas as pd
import glob
import re

Question 1: Load the data.

Download this CSV file with local twitter data.

Load the data into a DataFrame called tweets.

Name the columns: polarity, id, date, query, user, and tweet.

Store the values in the table using the ISO-8859-1 string encoding.

Display the first five rows of the DataFrame.

How many entries are in the table?

In [2]:
tweets = pd.read_csv('sentiment.csv', encoding='ISO-8859-1')


tweets.columns = ['polarity', 'id', 'date', 'query', 'user','tweet']

tweets.head()

polarity          id                          date     query  \
0         0  1467810672  Mon Apr 06 22:19:49 PDT 2009  NO_QUERY   
1         0  1467810917  Mon Apr 06 22:19:53 PDT 2009  NO_QUERY   
2         0  1467811184  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
3         0  1467811193  Mon Apr 06 22:19:57 PDT 2009  NO_QUERY   
4         0  1467811372  Mon Apr 06 22:20:00 PDT 2009  NO_QUERY   

            user                                              tweet  
0  scotthamilton  is upset that he can't update his Facebook by ...  
1       mattycus  @Kenichan I dived many times for the ball. Man...  
2        ElleCTF    my whole body feels itchy and like its on fire   
3         Karoli  @nationwideclass no, it's not behaving at all....  
4       joy_wolf                      @Kwesidei not the whole crew

In [3]:
len(tweets)

1599999

There are total 1599999 entries in the table.

Question 2: Sample twitter data.
Randomly sample 1% of the data and save to a new DataFrame called sample_tweets.

Display the first ten rows of the DataFrame.

How many entries are in the table?

In [4]:
sample_tweets = tweets.sample(n=16000)

sample_tweets.head(10)

polarity          id                          date     query  \
620722          0  2228512321  Thu Jun 18 14:31:58 PDT 2009  NO_QUERY   
1586623         4  2190833360  Tue Jun 16 03:51:11 PDT 2009  NO_QUERY   
531250          0  2196254501  Tue Jun 16 12:13:43 PDT 2009  NO_QUERY   
1173467         4  1980770400  Sun May 31 07:37:48 PDT 2009  NO_QUERY   
1184690         4  1982650488  Sun May 31 11:37:54 PDT 2009  NO_QUERY   
771849          0  2302530534  Tue Jun 23 16:56:26 PDT 2009  NO_QUERY   
83713           0  1753422718  Sun May 10 01:33:46 PDT 2009  NO_QUERY   
289176          0  1994869677  Mon Jun 01 12:35:12 PDT 2009  NO_QUERY   
1495896         4  2070003327  Sun Jun 07 16:55:40 PDT 2009  NO_QUERY   
1120272         4  1973915473  Sat May 30 12:04:57 PDT 2009  NO_QUERY   

                    user                                              tweet  
620722        mutdlegend  @AndrewThomas89 sorry had to end bp Thursday e...  
1586623        Gay_Burns  @rudedoodle my fav balamory has to be the one ...  
531250          rbmartin  @felicityfuller Not you as well! @ms_cornwall'...  
1173467            jazzt  @thenewpsies  http://justjared.buzznet.com/ go...  
1184690     victoria8981  @friplaylist ... and 'Rock me Like a Hurricane...  
771849           fishaay  im actually so depressed im gonna go jump of d...  
83713            akneiva            im new on twitter , noone reads me yet   
289176      BreathingNow  Mu network-connectiojn has been down for a while   
1495896        emailelle  @afrobella your struggle = my struggle!  think...  
1120272  successfulwomen  Brilliant performance by #SusanBoyle well done...

In [5]:
len(sample_tweets)

16000

There are 16000 entries in this table.

Question 3: Drop unused data.
Drop id, date, query, and user columns from sample_tweets.

Display the first ten lines of the DataFrame.

In [6]:
sample_tweets.drop(['id', 'date','query', 'user'], axis = 1, inplace=True)

sample_tweets.head(10)

polarity                                              tweet
620722          0  @AndrewThomas89 sorry had to end bp Thursday e...
1586623         4  @rudedoodle my fav balamory has to be the one ...
531250          0  @felicityfuller Not you as well! @ms_cornwall'...
1173467         4  @thenewpsies  http://justjared.buzznet.com/ go...
1184690         4  @friplaylist ... and 'Rock me Like a Hurricane...
771849          0  im actually so depressed im gonna go jump of d...
83713           0            im new on twitter , noone reads me yet 
289176          0  Mu network-connectiojn has been down for a while 
1495896         4  @afrobella your struggle = my struggle!  think...
1120272         4  Brilliant performance by #SusanBoyle well done...

Question 4: Change the positive polarities.
Change all 4 polarity values to 1.
      Don't use a loop to make the changes. Use a lambda function.

Display the first ten lines of the DataFrame.

In [28]:
sample_tweets['polarity'] = sample_tweets['polarity'].apply(lambda x: 1 if x is 4 else x)


sample_tweets

polarity                                              tweet  \
620722          0  @AndrewThomas89 sorry had to end bp Thursday e...   
1586623         1  @rudedoodle my fav balamory has to be the one ...   
531250          0  @felicityfuller Not you as well! @ms_cornwall'...   
1173467         1  @thenewpsies  http://justjared.buzznet.com/ go...   
1184690         1  @friplaylist ... and 'Rock me Like a Hurricane...   
...           ...                                                ...   
683686          0  candy riot at the jamjar tonight. 3 bands +lot...   
1107570         1  Relaxing after another luverly ride....lots of...   
1511917         1  finally done with everything! re-reading eclip...   
316580          0                            working overtime again    
846221          1                                        brown hair    

                                                     split  caps  \
620722   [@AndrewThomas89, sorry, had, to, end, bp, Thu...    []   
1586623  [@rudedoodle, my, fav, balamory, has, to, be, ...  [PC]   
531250   [@felicityfuller, Not, you, as, well!, @ms_cor...    []   
1173467  [@thenewpsies, http://justjared.buzznet.com/, ...    []   
1184690  [@friplaylist, ..., and, 'Rock, me, Like, a, H...    []   
...                                                    ...   ...   
683686   [candy, riot, at, the, jamjar, tonight., 3, ba...    []   
1107570  [Relaxing, after, another, luverly, ride....lo...    []   
1511917  [finally, done, with, everything!, re-reading,...    []   
316580                          [working, overtime, again]    []   
846221                                       [brown, hair]    []   

                                                 lowercase hashtags  \
620722   [sorry, had, to, end, bp, early, had, to, some...       []   
1586623  [@rudedoodle, my, fav, balamory, has, to, be, ...       []   
531250   [@felicityfuller, you, as, well!, @ms_cornwall...       []   
1173467  [@thenewpsies, http://justjared.buzznet.com/, ...       []   
1184690        [@friplaylist, and, me, a, is, also, great]       []   
...                                                    ...      ...   
683686   [candy, riot, at, the, jamjar, tonight., bands...       []   
1107570  [after, another, luverly, ride....lots, of, cl...       []   
1511917  [finally, done, with, everything!, re-reading,...       []   
316580                          [working, overtime, again]       []   
846221                                       [brown, hair]       []   

                       mentions                             urls  
620722        [@AndrewThomas89]                               []  
1586623           [@rudedoodle]                               []  
531250   [@felicityfuller, @ms]                               []  
1173467          [@thenewpsies]  [http://justjared.buzznet.com/]  
1184690          [@friplaylist]                               []  
...                         ...                              ...  
683686                       []                               []  
1107570                      []                               []  
1511917                      []                               []  
316580                       []                               []  
846221                       []                               []  

[16000 rows x 8 columns]

Question 5: Count number of entries with each polarity.
How many entries are there for each polarity?

Use the GroupBy methods to find the answer.

In [8]:
sample_tweets.groupby('polarity').size()

polarity
0    7975
4    8025
dtype: int64

Question 6: Add the following columns to sample_tweets.
Create a new column called split that contains a list of the words in the tweet.

Create a new column called caps that contains a list of the words in the tweet that are all-in-caps.

For instance, HELLO.

For tweets without any words that are all-in-caps, the entry should be NaN.

Create a new column called hashtags that contains a list of the words in the tweet with a hashtag, the # symbol.

For tweets without any hashtags, the entry should be NaN.

Create a new column called mentions that contains a list of the words in the tweet that have mentions.

Mentions begin with the @ symbol.

For tweets without any mentions, the entry should be NaN.

Create a new column called urls with a list of the Uniform Resource Locators (URLs) in the tweet.

All URLS begin with the prefix http.

For tweets without any URLs, the entry should be NaN.

Create a new column called numbers that contains a list of the numbers, like 55, in the tweet.

For tweets without any numbers, the entry should be NaN.

Create a new column called lowercase that contains a list of the words in the original tweet in all lowercase.
Note: You should use the pandas apply method along with lambda functions to create these lists.

In [18]:
sample_tweets['split']=sample_tweets['tweet'].apply(lambda x: x.split())

In [19]:
sample_tweets['caps']= sample_tweets.split.apply(lambda x: [i for i in x if i.isupper()])
sample_tweets['caps'].apply(lambda x: x if not len(x)==0 else np.nan)

620722      NaN
1586623    [PC]
531250      NaN
1173467     NaN
1184690     NaN
           ... 
683686      NaN
1107570     NaN
1511917     NaN
316580      NaN
846221      NaN
Name: caps, Length: 16000, dtype: object

In [42]:
sample_tweets['hashtags'] = sample_tweets['tweet'].apply(lambda x: re.findall(r"#(\w+)", x))
sample_tweets['hashtags'].apply(lambda x: x if not len(x)==0 else np.nan)

620722     NaN
1586623    NaN
531250     NaN
1173467    NaN
1184690    NaN
          ... 
683686     NaN
1107570    NaN
1511917    NaN
316580     NaN
846221     NaN
Name: hashtags, Length: 16000, dtype: object

In [43]:
sample_tweets['mentions']=sample_tweets['tweet'].apply(lambda x: re.findall(r'@[A-Za-z0-9]+',x))

sample_tweets['mentions'].apply(lambda x: x if not len(x)==0 else np.nan)

620722          [@AndrewThomas89]
1586623             [@rudedoodle]
531250     [@felicityfuller, @ms]
1173467            [@thenewpsies]
1184690            [@friplaylist]
                    ...          
683686                        NaN
1107570                       NaN
1511917                       NaN
316580                        NaN
846221                        NaN
Name: mentions, Length: 16000, dtype: object

In [44]:
sample_tweets['urls']=sample_tweets['tweet'].apply(lambda x: re.findall(r'https?://[A-Za-z0-9./]+',x))
sample_tweets['urls'].apply(lambda x: x if not len(x)==0 else np.nan)

620722                                 NaN
1586623                                NaN
531250                                 NaN
1173467    [http://justjared.buzznet.com/]
1184690                                NaN
                        ...               
683686                                 NaN
1107570                                NaN
1511917                                NaN
316580                                 NaN
846221                                 NaN
Name: urls, Length: 16000, dtype: object

In [23]:
sample_tweets['lowercase']= sample_tweets.split.apply(lambda x: [i for i in x if i.islower()])

Question 7: Create a string of the stems of all of the words in the tweets. In linguistics, a stem is the part of a word responsible for its lexical meaning. Stemming is the act of taking a word and reducing it into its stem. For some words, its stem is a stand-alone word. For instance, the stem for the word writing is write. But a stem isn't always a stand-alone word. For example, the stem of the words study, studies, and studying is studi.

The NLTK is a popular platform for building Python programs to work with human language data. It includes a Python package for stemming words. To use this package, your Jupyter notebook should include the following import statements:

import nltk from nltk.stem import PorterStemmer

To access the functionality of this package, you first have to create a new PorterStemmer object, like this:

ps=PorterStemmer()

Then, you can use the stem method of this object to find the stem of a given word. For instance, to find the stem for the word writing, you would use this code:

ps.stem('writing')

which should return the stem 'write'.

Create a new column called stem.

Using the lowercase list of words of each tweet, apply the stem method to each word and create a list of all of the stem words.

Use a lambda function to access each row, and within the lambda function, use a loop to find the stem of each word.

Combine of the list of stems into a single string and store in the stem column.

Use the join method in the string class to create a single string.

Each word should be separated by a single space.

In [34]:
from nltk.stem.porter import *
ps = PorterStemmer()

sample_tweets['stem'] = sample_tweets['tweet'].apply(lambda x: [ps.stem(i) for i in x])
for i in range(len(stem)):
    stem[i] = ''.join(stem[i])

Question 8: Save sample_tweets to a file. Dump your dataframe to a file called hw04.csv

In [46]:
sample_tweets.to_csv('hw04.csv',encoding='utf-8')